File: server/Schema/CSVData/CSVData_IDAndCount.cs
//////////////////////////////////////////////////////////////////////////
//
// 文件：Schema/CSVData/CSVData_IDAndCount.cs
// 作者：Xoen Xie
// 时间：2023/06/19
// 描述：CSV数据 id count组
// 说明：
//
//////////////////////////////////////////////////////////////////////////
using System;
using XSF;

namespace XsfScp
{
    public struct CSVIdCount
    {
        public uint id;
        public uint count;
    }

    public sealed class CSVData_IDAndCount : CSVData
    {
        public override CSVDataType DataType { get { return CSVDataType.IDAndCount; } }

        public CSVIdCount[]? icValue;

        public override void Read(int nRow, int nCol, string sData)
        {
            if (string.IsNullOrEmpty(sData))
            {
                icValue = null;
                return;
            }


            try
            {
                string[] item = sData.Split('|');
                icValue = new CSVIdCount[item.Length];
                for (int i = 0; i < item.Length; ++i)
                {
                    string[] itemData = item[i].Split(':');
                    if (itemData.Length != 2)
                    {
                        throw new XSFSchemaLoadException ($"CSVData_IDAndCount.Read error, should be id:count, str={sData}, row={nRow}, col={nCol}");
                    }

                    CSVIdCount ic;
                    ic.id = Convert.ToUInt32(itemData[0]);
                    ic.count = Convert.ToUInt32(itemData[1]);
                    icValue[i] = ic;
                }
            }
            catch (Exception e)
            {
                Serilog.Log.Error($"CSVData_IDAndCount.Read error str={sData}, row={nRow}, col={nCol}, message={e.Message}");
                throw;
            }
        }

    }
}

File: server-cpp/source/schema/src/csv_data/CSVData_IDAndCount.cpp
//////////////////////////////////////////////////////////////////////////
//
// 文件：source/xsf/src/schema/csv_data/CSVData_IDAndCount.cpp
// 作者：Xoen Xie
// 时间：2023/08/14
// 描述：CSV数据 id count组
// 说明：
//
//////////////////////////////////////////////////////////////////////////

#include "CSVData_IDAndCount.h"
#include "XSFMallocHook.h"
#include "XSF.h"

namespace xsf_scp
{
    void CSVIdCountArray::Release()
    {
        nTotal = 0;
        XSF_FREE(pDatas);
    }

    bool CSVData_IDAndCount::Read(uint32 row, uint32 col, const char * sData)
    {
        icValue.nTotal = 0;
        icValue.pDatas = nullptr;

        if(strlen(sData) <= 0)
            return true;

        vector<string> datas;
        XSFCore::Split(sData, '|', datas);

        icValue.nTotal = datas.size();
        icValue.pDatas = (CSVIdCount*)xsf_malloc(sizeof(CSVIdCount)*icValue.nTotal);
        for(uint i = 0; i < icValue.nTotal; i ++)
        {
            vector<string> idc;
            XSFCore::Split(datas[i].c_str(), ':', idc);
            if(idc.size() != 2)
            {
                XSF_ERROR("CSVData_IDAndCount::Read idc.size() != 2, row=%d, col=%d, data=%s, idc=%s", row, col, sData, datas[i].c_str());
                return false;
            }

            icValue.pDatas[i].nID = strtoul(idc[0].c_str(), 0, 10);
            icValue.pDatas[i].nCount = strtoul(idc[1].c_str(), 0, 10);
        }

        return true;
    }
}
